File: TF_Study_PythonCode/9/9.1.4.py
from time import time
import tensorflow as tf
import numpy as np

# import tensorflow.examples.tutorials.mnist.input_data as input_data
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets("../data/MNIST_data/", one_hot=True)

# 输入层
x = tf.placeholder(tf.float32, [None, 784], name="x")
y = tf.placeholder(tf.float32, [None, 10], name="y")

# 隐藏层（隐含层）
H1_NN = 256
W1 = tf.Variable(tf.random_normal([784, H1_NN]))
b1 = tf.Variable(tf.zeros([H1_NN]))

Y1 = tf.nn.relu(tf.matmul(x, W1) + b1)  # 隐藏层的激活函数

# 输出层
W2 = tf.Variable(tf.random_normal([H1_NN, 10]))
b2 = tf.Variable(tf.zeros([10]))

forward = tf.matmul(Y1, W2) + b2
pred = tf.nn.softmax(forward)

# 训练参数
train_epochs = 40
batch_size = 50
total_batch = int(mnist.train.num_examples/batch_size)
display_step = 1
learning_rate = 0.01

# 交叉熵损失函数
loss_function = tf.reduce_mean(
    tf.nn.softmax_cross_entropy_with_logits(logits=forward, labels=y))

# Adam 优化器
optimizer = tf.train.AdamOptimizer(learning_rate).minimize(loss_function)

# 定义准确率
correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(pred, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

# 初始化开始训练的时间
startTime = time()

sess = tf.Session()
init = tf.global_variables_initializer()
sess.run(init)

for epoch in range(train_epochs):
    for batch in range(total_batch):
        xs, ys = mnist.train.next_batch(batch_size)
        sess.run(optimizer, feed_dict={x: xs, y: ys})

    # 完成一个批次后，计算误差与准确率
    loss, acc = sess.run([loss_function, accuracy],
                         feed_dict={x: mnist.validation.images,
                                    y: mnist.validation.labels})

    if(epoch+1) % display_step == 0:
        print("train epoch:", epoch+1, ' Loss=', loss, " accuracy=", acc)

# 评估模型
accu_test = sess.run(accuracy, feed_dict={
                     x: mnist.test.images, y: mnist.test.labels})
print("准确率：", accu_test)

prediction_result = sess.run(
    tf.argmax(pred, 1), feed_dict={x: mnist.test.images})

print('预测结果：', prediction_result[0:10])

compare_lists = prediction_result == np.argmax(mnist.test.labels, 1)
print('预测与实际的结果比较：', compare_lists)

err_lists = [i for i in range(len(compare_lists)) if compare_lists[i] == False]
print('错误数量：', len(err_lists), "   错误内容：", err_lists)


def print_predict_errs(labels, prediction):
    count = 0
    compare_lists = (prediction == np.argmax(labels, 1))
    err_lists = [i for i in range(
        len(compare_lists)) if compare_lists[i] == False]
    for x in err_lists:
        print("index=", str(x), "标签值=", np.argmax(
            labels[x]), " 预测值=", prediction[x])
        count = count+1
    print("总计：", str(count))

print_predict_errs(labels=mnist.test.labels, prediction=prediction_result)

# 以下为扩展方法

import matplotlib.pyplot as plt

def plot_images_labels_prediction(images,
                                  labels,
                                  prediction,
                                  index,
                                  num=10):
    fig = plt.gcf()
    fig.set_size_inches(10, 12)
    if num > 25:
        num = 25    # 最多显示25个子图
    for i in range(0, num):
        ax = plt.subplot(5, 5, i+1) # 获取要处理图像的位置
        ax.imshow(np.reshape(images[index], (28, 28)), cmap='binary') # 显示第index个图像
        title = "label=" + str(np.argmax(labels[index]))
        if len(prediction) > 0:
            title += ", predict=" + str(prediction[index])
        
        ax.set_title(title, fontsize=10)
        ax.set_xticks([])   # 不显示坐标轴
        ax.set_yticks([])
        index += 1
    
    plt.show()

err_images=[]
err_labels=[]
err_prediction=[]

for x in err_lists:
    err_images.append(mnist.test.images[x])
    err_labels.append(mnist.test.labels[x])
    err_prediction.append(prediction_result[x])

plot_images_labels_prediction(err_images, err_labels, err_prediction, 0, 25)

File: MOOC_TF_Study/_9/9_1_4.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using MOOC_TF_Study.Common;
using NumSharp;
using Tensorflow;
using static Tensorflow.Binding;
using Tensorflow.Hub;

namespace MOOC_TF_Study._9
{
    class _9_1_4
    {
        public static void Run()
        {
            fun1();
        }

        private static void fun1()
        {
            var path = "../../../../data/MNIST_data/";
            // 这里需要引用 Tensorflow.Hub 项目
            var mnist = MnistModelLoader.LoadAsync(path, oneHot:true).Result;

            // 输入层
            var x = tf.placeholder(TF_DataType.TF_DOUBLE, (-1, 784), name: "x");
            var y = tf.placeholder(TF_DataType.TF_DOUBLE, (-1, 10), name: "y");

            // 隐含层
            var H1_NN = 256;

            var w1 = tf.Variable(tf.random_normal((784, H1_NN), dtype:TF_DataType.TF_DOUBLE));
            var b1 = tf.Variable(tf.zeros((H1_NN), dtype: TF_DataType.TF_DOUBLE));

            var y1 = tf.nn.relu(tf.matmul(x, w1) + b1);

            // 输出层
            var w2 = tf.Variable(tf.random_normal((H1_NN, 10), dtype: TF_DataType.TF_DOUBLE));
            var b2 = tf.Variable(tf.zeros((10), dtype: TF_DataType.TF_DOUBLE));

            var forward = tf.matmul(y1, w2) + b2;
            var pred = tf.nn.softmax(forward);

            var train_epochs = 50;
            var batch_size = 100;
            var total_batch = (int) mnist.Train.NumOfExamples / batch_size;
            var display_step = 1;
            var learning_rate = 0.01f;

            // var loss_function = tf.reduce_mean(-tf.reduce_sum(y * tf.log(pred), reduction_indices: 1));
            var loss_function = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels: y, logits: forward));
            var optimizer = tf.train.GradientDescentOptimizer(learning_rate).minimize(loss_function);
            // var optimizer = tf.train.AdamOptimizer(learning_rate, TF_DataType.TF_DOUBLE).minimize(loss_function);

            var correct_prediction = tf.equal(tf.argmax(pred, 1), tf.argmax(y, 1));
            var accuracy = tf.reduce_mean(tf.cast(correct_prediction, TF_DataType.TF_DOUBLE));

            using (var sess = tf.Session())
            {
                var init = tf.global_variables_initializer();
                sess.run(init);

                for (var epoch = 0; epoch < train_epochs; epoch++)
                {
                    for (var batch = 0; batch < total_batch; batch++)
                    {
                        var (xs, ys) = mnist.Train.GetNextBatch(batch_size);
                        sess.run(optimizer, new FeedItem(x, xs), new FeedItem(y, ys));
                    }

                    var (loss, acc) = sess.run((loss_function, accuracy),
                        new FeedItem(x, mnist.Validation.Data),
                        new FeedItem(y, mnist.Validation.Labels));

                    if ((epoch + 1) % display_step == 0)
                        Console.WriteLine($"train epoch:{epoch + 1}   loss={loss} accuracy={acc}");
                }

                // 评估模型
                var accu_test = sess.run(accuracy, (x, mnist.Test.Data), (y, mnist.Test.Labels));
                Console.WriteLine("准确率：{0}", accu_test.ToString());

                var prediction_result = sess.run(tf.argmax(pred, 1), new FeedItem(x, mnist.Test.Data));
                Console.WriteLine($"测试结果：{prediction_result["0:10"]}");

                var compare_lists = prediction_result == np.argmax(mnist.Test.Labels, 1);
                Console.WriteLine("预测与实际结果的比较 {0}", compare_lists);

                var compareResult = (bool[])compare_lists;

                /*
                var err_list = new List<int>();
                for(var i = 0;i < compareResult.Length; i++)
                {
                    if (!compareResult[i])
                        err_list.Add(i);
                }
                */

                var err_list = compareResult.Select((result, index) => (result, index))
                                            .Where(o => !o.result)
                                            .Select(o => o.index).ToArray();

                var labels = mnist.Test.Labels;
                var prediction = (long[])prediction_result;
                foreach (var errIndex in err_list)
                {
                    Console.WriteLine($"index={errIndex} 标签值={np.argmax(labels[errIndex])} 预测值={prediction[errIndex]}");
                }
            }
        }
    }
}
